### ISSUE TITLE
Epic: Fix Lens Distortion Calibration Pipeline - RMSE Validation Discrepancy

### ISSUE THREAD

**Issue by Nuno-Monteiro** [GoCleverOrg/poc-homography#201]
## Summary
Fix the lens distortion calibration pipeline that uses parking lot lines instead of checkerboard patterns. The current implementation reports RMSE of ~0.5 pixels during coefficient calculation, but the distortion validator cannot verify this improvement due to 6 interconnected root causes spanning data collection, solver validation, intrinsics consistency, data loading, end-to-end verification, and metric interpretation.

## Context
The parking lot line-based calibration pipeline was designed as an alternative to traditional checkerboard calibration. It works by: (1) GeoTIFF line picker creates ground truth lines, (2) camera survey captures overlapping frames, (3) camera line annotator creates N-point edge pixel annotations, (4) lens_calibration webapp calculates coefficients via L-BFGS-B optimization, (5) distortion_validator webapp applies coefficients to measure straightness.

During testing, the solver reports RMSE of ~0.5 pixels after optimization, suggesting successful calibration. However, when the distortion_validator attempts to verify the improvement, it cannot demonstrate that the calculated coefficients actually straighten the lines. This indicates a fundamental breakdown in the pipeline - either the coefficients are not truly improving distortion correction, or the validation process is not correctly measuring the improvement.

Root cause analysis identified 6 interconnected issues that prevent successful end-to-end calibration and validation, ranging from insufficient sample points and missing pre-flight validation to intrinsics mismatches and silent data loading failures.

### In Scope
- **Issue 1: Insufficient Sample Points** - Address under-constrained optimization due to 6-15 points per line (checkerboard uses hundreds)
- **Issue 2: No Pre-flight Validation** - Implement validation for minimum points per line, total lines, line coverage, and orientation diversity before solver runs
- **Issue 3: Intrinsics Mismatch Potential** - Ensure consistent intrinsics between calibration and validation (currently both use defaults: fx=1000, fy=1000, cx=960, cy=540)
- **Issue 4: CameraLine.sample_points() Silent Fallback** - Fix silent fallback to LINEAR interpolation when edge_pixels is None, which masks data loading errors
- **Issue 5: End-to-End Data Flow Not Validated** - Add verification that edge_pixels from JSON load correctly into CameraLine and debug visualization of optimization data
- **Issue 6: RMSE Metric Misleading** - Distinguish in-sample (optimization) vs out-of-sample (validation) error to confirm generalization
- Establish minimum data requirements (points per line, number of lines, coverage, diversity)
- Create end-to-end integration test proving pipeline produces valid coefficients
- Document complete pipeline workflow with validation checkpoints

### Out of Scope
- Changing fundamental approach from parking lines to checkerboard patterns
- Complete UI redesigns of existing webapps
- Real-time camera integration (still using static survey images)
- Support for cameras beyond current Hikvision PTZ model
- Automatic line detection (still using manual N-point annotation)
- Multi-zoom calibration workflows (focus on single zoom factor)

## Definition of Done
- [ ] All 6 root causes have documented fixes or mitigations
- [ ] Pre-flight validation rejects insufficient data before optimization runs
- [ ] Intrinsics are explicitly configured and consistent across calibration and validation
- [ ] CameraLine.sample_points() raises error when edge_pixels is None instead of silent fallback
- [ ] End-to-end test demonstrates: annotate lines → calculate coefficients → validate improvement with validator webapp
- [ ] Validation test uses separate hold-out lines not used during calibration
- [ ] RMSE reported by solver is labeled "training RMSE" and RMSE from validator is labeled "validation RMSE"
- [ ] Minimum data requirements documented: minimum points per line, minimum total lines, coverage criteria, orientation diversity
- [ ] Pipeline produces coefficients where validation RMSE is measurably lower than baseline (uncorrected) RMSE
- [ ] Debug logging confirms edge_pixels data loads correctly from JSON into CameraLine objects

### Issue 1: Insufficient Sample Points for Mathematical Constraints

**Current State**: Camera line annotator produces 6-15 points per line using N-point mode. L-BFGS-B optimizer in `distortion_solver.py` uses these points to find 5 distortion coefficients (k1, k2, k3, p1, p2).

**Problem**: Traditional checkerboard calibration uses hundreds of corner points across dozens of frames. With 6-15 points per line, the optimization is severely under-constrained. The solver can find coefficients that minimize error on this sparse data but don't generalize to actual distortion correction.

**Evidence from Code**:
```python
# distortion_solver.py:48
num_samples_per_line: int = 20
```
This config default of 20 samples per line only matters if edge_pixels has enough data. When edge_pixels has 6-15 actual edge detections, the solver uses those sparse points directly.

**Requirements**:
- Establish minimum acceptable points per line based on mathematical constraints of 5-coefficient optimization
- Determine minimum number of lines needed for robust calibration
- Consider whether current annotation workflow can produce sufficient data density, or if automated edge detection is needed
- Document trade-off: more manual annotation effort vs calibration quality

**Reference**: Checkerboard calibration typically uses 200-400 corner points across 10-20 frames. Zhang's calibration method (OpenCV calibrateCamera) requires minimum 4 views with 6+ points each, but achieves accuracy with 10+ views of 20+ points.

### Issue 2: No Pre-flight Validation

**Current State**: The solver in `distortion_solver.py:130-154` accepts any input as long as `len(lines) >= 1` and intrinsic matrix is 3x3. No validation of data quality.

**Problem**: Solver runs with insufficient data and returns "successful" optimization results that are mathematically valid but meaningless for actual calibration.

**Evidence from Code**:
```python
# distortion_solver.py:149-153
if not lines:
    raise ValueError("At least one line required for calibration")

if intrinsic_matrix.shape != (3, 3):
    raise ValueError(f"Intrinsic matrix must be 3x3, got {intrinsic_matrix.shape}")
```
Only checks for non-empty list and correct matrix shape. No validation of points per line, line diversity, or coverage.

**Requirements**:
- Minimum points per line (reject lines with fewer than N points where N is determined from Issue 1 analysis)
- Minimum total lines (reject datasets with fewer than M lines)
- Line coverage across image (reject if all lines are in one corner - need spatial distribution)
- Line orientation diversity (reject if all lines are parallel - need at least 2 different orientations)
- Add validation function that runs before optimization with clear error messages explaining what's missing

**Reference**: OpenCV calibrateCamera has minimum requirements but also quality metrics. Emulate this pattern for parking line calibration.

### Issue 3: Intrinsics Mismatch Potential

**Current State**: Both `lens_calibration/views.py` and `distortion_validator/views.py` use default intrinsics if not provided:
```python
# lens_calibration/views.py:121-127
intrinsic_matrix = np.array([
    [intrinsics.get("fx", 1000.0), 0.0, intrinsics.get("cx", 960.0)],
    [0.0, intrinsics.get("fy", 1000.0), intrinsics.get("cy", 540.0)],
    [0.0, 0.0, 1.0],
])
```

**Problem**: If calibration uses defaults (fx=1000, fy=1000, cx=960, cy=540) but validator uses different defaults or image-dependent values (e.g., cx=w/2, cy=h/2), the undistortion math will be inconsistent even with identical distortion coefficients.

**Evidence from Code**:
```python
# distortion_validator/views.py:323-326
fx = intrinsics.get("fx", 1000.0)  # Default focal length
fy = intrinsics.get("fy", 1000.0)
cx = intrinsics.get("cx", w / 2)   # Default to image center
cy = intrinsics.get("cy", h / 2)
```
Validator defaults cx/cy to image center (width/2, height/2), while calibrator defaults to 960, 540. For a 1920x1080 image, validator uses cx=960, cy=540 (matches), but for other resolutions this creates mismatch.

**Requirements**:
- Intrinsics must be explicitly configured in calibration workflow (no silent defaults)
- Validator must use exact same intrinsics as calibrator (read from calibration result or shared config)
- Add validation that intrinsics match between calibration and validation
- Consider storing intrinsics alongside distortion coefficients in calibration output
- Add warning logs when defaults are used

**Context**: The Brown-Conrady distortion model is sensitive to principal point (cx, cy) location. Even small differences (10-20 pixels) in principal point can cause significant errors in undistortion, especially near image edges where distortion is most severe.

### Issue 4: CameraLine.sample_points() Silent Fallback

**Current State**: The `models.py:94-121` implementation silently falls back to linear interpolation if edge_pixels is None:

**Evidence from Code**:
```python
# models.py:107-121
if self.edge_pixels is not None and len(self.edge_pixels) >= 2:
    # Use actual edge pixels - these contain the distortion signal
    edge_arr = np.array(self.edge_pixels, dtype=np.float64)
    if len(edge_arr) <= num_samples:
        return edge_arr
    # Subsample evenly
    indices = np.linspace(0, len(edge_arr) - 1, num_samples, dtype=int)
    return edge_arr[indices]

# Fallback to linear interpolation (only useful for visualization)
t = np.linspace(0, 1, num_samples)
start = np.array(self.start_pixel)
end = np.array(self.end_pixel)
points = start + np.outer(t, end - start)
return points.astype(np.float64)
```

**Problem**: Linear interpolation between start_pixel and end_pixel produces perfectly straight lines with zero distortion signal. If edge_pixels fails to load from JSON (None), the solver optimizes on straight lines and finds zero distortion (correct for the data, wrong for reality). This masks data loading errors silently - no error raised, just meaningless results.

**Requirements**:
- Change sample_points() to raise ValueError if edge_pixels is None when called by calibration code
- Add explicit mode parameter: `sample_points(num_samples, mode="edge_pixels"|"linear")`
- Calibration code must explicitly request "edge_pixels" mode and handle the error if data is missing
- Visualization code can use "linear" mode for displaying start/end endpoints
- Add logging to show when edge_pixels is used vs linear fallback

**Context**: The comment "only useful for visualization" indicates developer knew this was a problem, but the silent fallback creates a dangerous failure mode where pipeline appears to work but produces garbage coefficients.

### Issue 5: End-to-End Data Flow Not Validated

**Current State**: No verification that edge_pixels from camera_line_annotator JSON export actually loads into CameraLine objects correctly. No debug output showing what data the optimizer is actually seeing.

**Problem**: Cannot diagnose whether failure is due to:
- JSON export from camera_line_annotator not including edge_pixels
- JSON import in lens_calibration not parsing edge_pixels correctly
- edge_pixels format mismatch (list vs tuple vs numpy array)
- Optimizer receiving wrong data structure

**Evidence from Code**:
```python
# lens_calibration/views.py:343-365
for cal_line in calibration_lines:
    points = cal_line.get("points", [])
    if len(points) < 2:
        continue

    line_id = cal_line.get("line_id") or f"line_{cal_line.get('index', 0):04d}"

    # Convert points to tuples for edge_pixels
    edge_pixels = tuple(tuple(pt) for pt in points)

    # Use first and last point as start/end
    start_pixel = tuple(points[0])
    end_pixel = tuple(points[-1])

    camera_line = CameraLine(
        line_id=f"{line_id}_{image_name}",
        image_path=image_name,
        start_pixel=start_pixel,
        end_pixel=end_pixel,
        ptz_position=ptz,
        confidence=1.0,
        edge_pixels=edge_pixels,
    )
```
This code converts points from JSON to edge_pixels, but there's no verification that edge_pixels is non-empty or that it successfully loads into the CameraLine object.

**Requirements**:
- Add validation after CameraLine construction: assert that edge_pixels is not None and len(edge_pixels) >= minimum
- Add debug logging showing: number of edge_pixels loaded, range of coordinates, line_id
- Add endpoint in lens_calibration to dump loaded CameraLine objects as JSON for inspection
- Add visualization overlay in validator webapp showing edge_pixels on image to confirm correct loading
- Create unit test that round-trips: JSON export → CameraLine construction → sample_points() → verify points match

**Reference**: The camera_line_annotator views.py shows JSON export format, but doesn't show the corresponding "calibration_lines" format with N-point data. Need to trace data format through entire pipeline.

### Issue 6: RMSE Metric Misleading (In-Sample vs Out-of-Sample)

**Current State**: The solver reports RMSE calculated on the same points used for optimization:

**Evidence from Code**:
```python
# distortion_solver.py:210-226
for i, samples in enumerate(line_samples):
    undistorted = self._undistort_points(samples, final_coeffs, cx, cy, fx, fy)
    error = self._line_straightness_error(undistorted)
    num_samples = len(samples)
    rmse = np.sqrt(error / num_samples) if num_samples > 0 else 0.0
    rmse_per_line.append(float(rmse))
```
This calculates RMSE on `line_samples`, which are the same points used during optimization. This is in-sample error, not validation error.

**Problem**: In-sample RMSE of 0.5 pixels tells you the optimizer successfully minimized the objective function on the training data. It does NOT tell you if the coefficients generalize to correct actual lens distortion. This is classic overfitting - optimizer finds coefficients that fit sparse noisy data perfectly but don't reflect true distortion model.

**Requirements**:
- Distinguish "training RMSE" (in-sample, calculated by solver) vs "validation RMSE" (out-of-sample, calculated on hold-out lines)
- Validation workflow must use lines NOT included in calibration dataset
- UI must clearly label these metrics: "Training RMSE (optimization result)" vs "Validation RMSE (generalization test)"
- Accept that training RMSE will be lower than validation RMSE (this is expected)
- Define success criterion: validation RMSE must be measurably better than baseline (uncorrected) RMSE

**Context**: Standard machine learning practice - never evaluate model performance on training data. Need separate validation set. For calibration, this means: annotate 2x lines, use half for calibration, half for validation.

### Minimum Data Requirements Recommendation

Based on mathematical constraints of 5-coefficient optimization and checkerboard calibration benchmarks:

**Recommended Minimums**:
- **Points per line**: 15 minimum, 30+ preferred (need to over-constrain 5 coefficients)
- **Total lines**: 10 minimum, 20+ preferred (need spatial diversity and validation hold-out)
- **Line coverage**: Lines must cover all 4 quadrants of image (distortion varies spatially)
- **Orientation diversity**: At least 3 different orientations (0°, 45°, 90° ideally) to constrain radial vs tangential distortion independently

**Rationale**:
- 5 coefficients to solve for (k1, k2, k3, p1, p2)
- Each point provides 2 constraints (x and y perpendicular distance to line)
- Minimum 15 points × 2 = 30 constraints per line
- With 10 lines × 30 = 300 total constraints >> 5 unknowns (60:1 ratio)
- This matches checkerboard ratio (200-400 points for 5 coefficients)

**Validation Split**:
- Use 50% of lines for calibration (training set)
- Use 50% of lines for validation (test set)
- Never optimize on validation lines

### Proposed Sub-Issues

This Epic decomposes into the following implementation issues (sub-issue files NOT created yet, but outlined here for planning):

**Sub-Issue 1**: "Add pre-flight validation to distortion solver"
- Implement validation function in distortion_solver.py
- Check minimum points per line, total lines, coverage, orientation diversity
- Raise clear errors before optimization runs
- Add configuration for minimum thresholds
- Estimated effort: 1-2 days

**Sub-Issue 2**: "Fix CameraLine.sample_points() silent fallback to raise error"
- Change models.py to raise ValueError when edge_pixels is None in calibration mode
- Add mode parameter to distinguish calibration vs visualization use
- Update all callers to handle new error or specify mode
- Add unit tests for error conditions
- Estimated effort: 1 day

**Sub-Issue 3**: "Ensure intrinsics consistency between calibration and validation"
- Add intrinsics to calibration result output (save alongside coefficients)
- Update validator to read intrinsics from calibration file
- Add validation check that intrinsics match
- Remove silent defaults (require explicit configuration)
- Add warning logs when defaults are used
- Estimated effort: 1-2 days

**Sub-Issue 4**: "Add end-to-end data flow validation and debug visualization"
- Add post-construction validation of edge_pixels in lens_calibration/views.py
- Add debug logging showing edge_pixels count, range, line_id
- Add endpoint to dump loaded CameraLine objects as JSON
- Add visualization in validator showing edge_pixels overlaid on image
- Create round-trip unit test: JSON → CameraLine → sample_points()
- Estimated effort: 2-3 days

**Sub-Issue 5**: "Separate training vs validation RMSE reporting"
- Update solver result to label RMSE as "training_rmse"
- Update validator to calculate "validation_rmse" on hold-out lines
- Update UI to clearly distinguish these metrics
- Update documentation explaining in-sample vs out-of-sample
- Estimated effort: 1 day

**Sub-Issue 6**: "Create end-to-end integration test with minimum data requirements"
- Document minimum data requirements in calibration docs
- Create test dataset meeting minimums (15+ points/line, 20+ lines, coverage, diversity)
- Implement test: annotate → calibrate (50% lines) → validate (50% lines)
- Assert validation RMSE < baseline RMSE (demonstrates improvement)
- Assert training RMSE < validation RMSE (normal overfitting pattern)
- Estimated effort: 2-3 days

**Total Estimated Effort**: 8-12 days (1.5-2 weeks) for one contributor

### Key Files
- `/Users/nuno.monteiro/Dev/SmartTerminal/poc-homography/.worktrees/feat/lens-distortion-calibration/poc_homography/calibration/lens_distortion/distortion_solver.py` - L-BFGS-B optimization solver
- `/Users/nuno.monteiro/Dev/SmartTerminal/poc-homography/.worktrees/feat/lens-distortion-calibration/poc_homography/calibration/lens_distortion/models.py` - CameraLine with edge_pixels and silent fallback
- `/Users/nuno.monteiro/Dev/SmartTerminal/poc-homography/.worktrees/feat/lens-distortion-calibration/webapp/lens_calibration/views.py` - Calibration API endpoint
- `/Users/nuno.monteiro/Dev/SmartTerminal/poc-homography/.worktrees/feat/lens-distortion-calibration/webapp/distortion_validator/views.py` - Validation API endpoint
- `/Users/nuno.monteiro/Dev/SmartTerminal/poc-homography/webapp/camera_line_annotator/views.py` - N-point annotation export

### Related Techniques
- Zhang's calibration method (OpenCV calibrateCamera) - establishes minimum data requirements for camera calibration
- Brown-Conrady distortion model - used in this pipeline, sensitive to principal point location
- L-BFGS-B optimization - current solver algorithm, requires well-constrained problem
- Total least squares (SVD line fitting) - used in straightness RMSE calculation

### Pipeline Workflow Diagram

```
1. GeoTIFF Line Picker → Cartografia_valencia_lines.yaml (ground truth)
2. Camera Survey → overlapping frames at different PTZ positions
3. Camera Line Annotator (N-point mode) → calibration_lines.json with edge_pixels
4. Lens Calibration Webapp → L-BFGS-B optimization → coefficients + training_rmse
5. Distortion Validator Webapp → apply coefficients → validation_rmse (hold-out lines)
6. Success: validation_rmse < baseline_rmse (demonstrable improvement)
```

**Validation Checkpoints** (to be added):
- Before Step 4: Pre-flight validation (minimum points, lines, coverage, diversity)
- After Step 3: Verify edge_pixels loaded correctly from JSON
- After Step 4: Verify intrinsics match between calibration and validation
- After Step 5: Compare training vs validation RMSE (expect validation > training)
